Framework: Update/abtest to filter non english

rralian opened the pull request:
We discovered that we are unintentionally including non-English users in our abtests when they are logged-out, which is the primary case for our NUX. This is an issue because most experiments will not have had the time for any new strings to be translated, which means users in other languages are getting partially translated pages, and that's going to hurt their experience and also hurt the performance of the test variant.

So this PR adds two checks for non-english locales. The first check is against `navigator.language`, which is the first preference in `navigator.languages` in all modern browsers except IE10. I'm using `navigator.userLanguage` as the fallback for IE10.

I am also adding a check against `i18nUtils.getLocaleSlug()`, which can be different from `navigator.language` if the user has gone to a page with a language in the url (e.g., https://wordpress.com/start/survey/es). This can happen with a direct link or if the user has accepted a language suggestion from the `locale-suggestions` component.

Last, there's quite a bit more changes in this PR because I addressed all linting issues and es6-ified the `abtest` module. You can view the individual commits to see these changes separately.

## Testing
First, you can run the tests in this branch and read through them to see what's been added.

```bsh
npm run test-client client/lib/abtest/test/index.js
```

### English Case -- Make Sure It Still Works
Open an incognito browser, set the debug string to `calypso:analytics`, open your inspector console and filter for `calypso_abtest_start" called` and go through the nux flow at `http://calypso.localhost:3000/start`. You should see several events get triggered as you go through the process.

![screen shot 2016-07-08 at 12 34 44 pm](https://cloud.githubusercontent.com/assets/134044/16696014/7c5a0bde-4508-11e6-8f14-710366927f24.png)

### Non-English as Primary Browser Language
Now go into your chrome preferences (use whatever browser you want, but these instructions are for chrome) and click "show advanced settings" on the bottom.

![screen shot 2016-07-08 at 10 03 28 am](https://cloud.githubusercontent.com/assets/134044/16696098/e6a081e4-4508-11e6-962e-88f60ebc028a.png)

Now go to Languages

![screen shot 2016-07-08 at 10 03 41 am](https://cloud.githubusercontent.com/assets/134044/16696106/f1eb30f8-4508-11e6-9b9e-95ce930ce30e.png)

And add a non-English language and drag it to the top.

![screen shot 2016-07-08 at 10 04 05 am](https://cloud.githubusercontent.com/assets/134044/16696121/053d8a5c-4509-11e6-8edc-1809cccad71d.png)

Close your incognito window and open a new one to clear out cookies and localStorage. Then load up http://calypso.localhost:3000. Clear your localStorage again via console. And now set the debug listener string to `calypso:analytics,calypso:abtests`.

Now if you go through the NUX you should no longer see the calypso_abtest_event firing.

### Non-English as OS Language
This one's a little tricker, but you can trigger this case by changing your entire OS language, which will require a restart.

### Non-English as Accepted Suggestion
Change your OS language back, and then go back into your browser preferences and set a non-English language as one of your preferences, but leave English as the _top_ preference. Now when you go through the NUX flow, you should get a suggestion to view the NUX in that language. Accept the suggestion and go through the NUX with the new language. You may have seen some `calypso_abtest_start` events prior to accepting the language change, but you should no longer see those events after switching languages.

/cc @coreh @michaeldcain @meremagee @gwwar @mtias @blowery 

Test live: https://calypso.live/?branch=update/abtest-filter-non-english

gwwar commented on client/lib/abtest/index.js:
```diff
@@ -1,25 +1,25 @@
-
+import activeTests from 'lib/abtest/active-tests';
+import analytics from 'lib/analytics';
+import i18n from 'i18n-calypso';
```
This is external now

rralian replied:
Right, thanks, I moved it.

gwwar commented:
I tested all three flows, and they work as advertised. With an operating system language set, not all browsers will update the window.Navigator values. 

For example with French set, FF returns:
<img width="138" alt="capture d ecran 2016-07-08 a 5 44 34 pm" src="https://cloud.githubusercontent.com/assets/1270189/16705042/18ecd23a-4534-11e6-8697-b324fb503a95.png">
While Chrome returns:
<img width="153" alt="capture d ecran 2016-07-08 a 5 44 26 pm" src="https://cloud.githubusercontent.com/assets/1270189/16705043/1b014d1c-4534-11e6-9beb-12236d8adfc6.png">

Let's wait for another 👍 on the code changes since this is a larger diff.

rralian commented:
> With an operating system language set, not all browsers will update the window.Navigator values.

You're right. I was actually surprised at chrome's behavior, tying `navigator.language` to the OS language, rather than just taking the first preference from `navigator.languages` like firefox does. I had initially thought `navigator.language` _always_ reflected the first option in `navigator.languages`, and updated my tests and documentation when I found that not to be the case.

In contrast, Safari doesn't have its own language setting separate from the OS. So it will reflect the OS language in `navigator.language` and it doesn't even support `navigator.languages`.

Opera works like chrome, `navigator.language` is the OS language, and `navigator.languages` reflects the application-specific language settings.

I've tested on IE11 and _none_ of the available language settings seem to reflect the language preferences I have set in Internet Options. I believe they all reflect the OS language, though it's not super-simple to switch OS languages in Windows using a test version in VirtualBox, so I'm not positive. I had added `navigator.userLanguage` because I _believe_ that's what IE10 uses to report the browser language setting, but I don't have a good way to check that.

I think checking both `navigator.language` and `navigator.languages[0]` is still the right thing to do in general, but maybe I should update the wording of the tests and the debug logs. Open to feedback/suggestions there.

deBhal commented on client/lib/abtest/index.js:
```diff
@@ -100,17 +134,34 @@ ABTest.prototype.getVariation = function() {
 ABTest.prototype.isEligibleForAbTest = function() {
-	var selectedSite = sites.getSelectedSite();
+	const selectedSite = sites.getSelectedSite();
+	const client = ( typeof navigator !== 'undefined' ) ? navigator : {};
```
We're only doing this to avoid a `ReferenceError`, so it might be clearer to shadow `navigator` rather than introducing a new variable.

rralian replied:
you mean like this? I suppose that is better.
```
const navigator = ( typeof navigator !== 'undefined' ) ? navigator : {};
```

deBhal replied:
Yeah, exactly

rralian replied:
Unfortunately our linting rules don't like that. I get the error `'navigator' was used before it was defined`, when I try that, which I do not get when I leave it as-is in the PR.

![lint error](https://cloudup.com/cuSgo_jRMqg+)

deBhal commented on client/lib/abtest/index.js:
```diff
@@ -100,17 +134,34 @@ ABTest.prototype.getVariation = function() {
+			debug( '%s: User has a non-English locale', this.experimentId );
+			return false;
+		}
+		if ( ! isUserSignedIn() && ! clientLocale.match( /^en\-?/i ) ) {
```
We don't need to escape the `-` in these regexps, do we?

rralian replied:
Nope, you're right. Updated and refactored to store the regex as a constant.

deBhal commented:
I'm not sure about giving the OS locale so much weight.  If a user's OS is in a non-English language, there's nothing they can do to be included.

If I set my OS to French, my browser settings (request headers) determine whether I end up at `wordpress.com` or `fr.wordpress.com` (or some third locale), but there's nothing I can do to be included in the ab tests.

Erring on the side of false negatives is much better than false positives here, and this seems like an edge case, so 👍  🚢 , but we might want to iterate again as well.

Our concern here is polluting the tests with untranslated content, so if the session locale is not English, that's definitely a `return false`, but after that I think the other tests could be more forgiving - if we've got `en` as the first element in `accept-language`, that seems like enough to let someone in.

Maybe we want to add something to i18n-calypso to differentiate between "We believe the user speaks English" and "We don't know, so we'll serve English".

rralian commented:
> I'm not sure about giving the OS locale so much weight.

@deBhal thanks for the reviews. I've found that browsers are surprisingly inconsistent in their treatment of these different values ([see this comment](https://github.com/Automattic/wp-calypso/pull/6632#issuecomment-231555878)), so browsing preference could be represented by either value. Because of this I think it's still appropriate to check each of these values.  However, I've updated the variable names, debug logs, and test wording to account for the fact that these variables do not necessarily have the same meaning across browsers.

rralian commented:
I've rebased, re-tested, and am merging this now.
